lib/task.js

```diff
@@ -351,10 +435,64 @@ Task.prototype = {
       pullOptions.authconfig = credentials;
     }
 
-    return yield pullImage(
+    return await pullImage(
       this.runtime.docker, dockerImageName, this.stream, pullOptions
     );
-  },
+  }
+
+  isAborted() {
+    return this.taskState === 'aborted';
+  }
+
+  isCanceled() {
+    return this.taskState === 'canceled';
+  }
+
+  /**
+   * Aborts the running of the task.  This is similar to cancelling a task, but
+   * will allow time to upload artifacts and report the run as an exception instead.
+   *
+   * @param {String} reason - Reason for aborting the test run (Example: worker-shutdown)
+   */
+  abort(reason='aborted') {
+    this.taskState = 'aborted';
+    this.taskException = reason;
+    this.runtime.stats.increment('tasks.aborted');
+    this.runtime.log('abort task', {
+      taskId: this.status.taskId,
+      runId: this.runId,
+      reason: reason
+    });
+
+    if (this.dockerProcess) { this.dockerProcess.kill(); }
+
+    this.stream.write(
+      this.fmtLog(`Task has been aborted prematurely. Reason: ${reason}`)
+    );
+  }
+
+  /**
+   * Cancel the running of the task.  Task cancellation was performed by an external
+   * entity and has already been published to task-exception exchange.  This will
+   * kill the docker container that might be running, attempt to release resources
+   * that were linked, as well as prevent artifacts from uploading, which cannot
+   * be done after a run is resolved.
+   *
+   * @param {String} reason - Reason for cancellation
+   */
+  cancel(reason) {
+    this.taskState = 'canceled';
+    this.taskException = reason;
```

Comment:
Again I don't see how reason makes any sense here.

If cancelled, you may not report state... Or at least you shouldn't it'll just give you an error.